sympy/assumptions/refine.py

```diff
@@ -393,6 +391,60 @@ def refine_matrixelement(expr, assumptions):
             return expr
         return MatrixElement(matrix, j, i)
 
+def refine_Add(expr, assumptions):
+    """
+    Handler for addition with infinities and NaN.
+
+    Detects special infinite values in a sum and refines the result accordingly.
+
+    Examples
+    ========
+
+    >>> from sympy import S, Add
+    >>> from sympy.assumptions.refine import refine_Add
+
+    >>> refine_Add(Add(S.NaN, 1, evaluate=False), {})
+    nan
+    """
+    finite_terms = []
+    inf = None
+
+    for arg in expr.args:
+        if arg is S.NaN:
+            return S.NaN
+
+        if arg in (S.Infinity, S.NegativeInfinity, S.ComplexInfinity) or \
+           (hasattr(arg, 'is_infinite') and arg.is_infinite):
+            if inf is None:
+                inf = arg
+            else:
+                inf_real, inf_imag = inf.as_real_imag()
+                arg_real, arg_imag = arg.as_real_imag()
+                if (inf_real.is_infinite and arg_real.is_infinite and inf_real != arg_real) or (inf_imag.is_infinite and arg_imag.is_infinite and inf_imag != arg_imag):
+                    return S.NaN
+                elif (arg_real.is_infinite or inf_real.is_infinite) and (arg_imag.is_infinite or inf_imag.is_infinite):
+                    return S.ComplexInfinity
```

Comment:
We also need to handle finite components here:

```suggestion
                    return S.ComplexInfinity
                elif inf_real.is_finite and arg_real.is_finite:
                    finite_terms.append(arg_real)
                elif inf_imag.is_finite and arg_imag.is_finite:
                    finite_terms.append(arg_imag * I)
```